### PR TITLE
Fixes false negatives in `UnnecessaryAbstractClass`

### PR DESCRIPTION
**Pull request by BraisGabin**
This PR fixes a false positive in `UnnecessaryAbstractClass`.

Before this PR the class `B` wasn't flagged:

```kotlin
interface A {
    val i: Int
}
abstract class B : A
```

And the same here:

```kotlin
abstract class A {
    abstract val i: Int
}
abstract class B : A()
```

This PR also improves the messages reported for this rule.

### CODE REVIEW

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/IndentingXMLStreamWriter.kt [hunk 1]
```diff
@@ -5,9 +5,9 @@ import javax.xml.stream.XMLStreamWriter
 
 @Suppress("TooManyFunctions")
 internal class IndentingXMLStreamWriter(
-    writer: XMLStreamWriter,
+    private val writer: XMLStreamWriter,
     private val indent: String = "  "
-) : DelegatingXMLStreamWriter(writer) {
+) : XMLStreamWriter by writer {
```
Comment (schalkms) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/IndentingXMLStreamWriter.kt:10]: Is there a reason why the two classes regarding the `XMLStreamWriter` are refactored and part of this PR?

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt [hunk 2]
```diff
@@ -73,12 +73,13 @@ class UnnecessaryAbstractClass(config: Config = Config.empty) : Rule(config) {
...
-            val body = klass.body
-            if (body != null) {
-                val namedMembers = body.children.filterIsInstance<KtNamedDeclaration>()
-                val namedClassMembers = NamedClassMembers(klass, namedMembers)
-                namedClassMembers.detectAbstractAndConcreteType()
-            } else if (klass.superTypeListEntries.isEmpty()) {
+            val namedMembers = klass.body?.children.orEmpty().filterIsInstance<KtNamedDeclaration>()
+            if (namedMembers.isNotEmpty()) {
+                NamedClassMembers(klass, namedMembers)
+                    .detectAbstractAndConcreteType()
```
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt:79]: The formatting looks a bit weird here IMHO.

```suggestion
                NamedClassMembers(klass, namedMembers).detectAbstractAndConcreteType()
```

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt [hunk 3]
```diff
@@ -73,12 +73,13 @@ class UnnecessaryAbstractClass(config: Config = Config.empty) : Rule(config) {
...
+            } else if (!klass.hasConstructorParameter()) {
+                report(CodeSmell(issue, Entity.from(klass), noConcreteMember), klass)
```
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt:81]: Where did you improve the message description?

> This PR also improves the messages reported for this rule.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt [hunk 4]
```diff
@@ -38,28 +33,56 @@ class UnnecessaryAbstractClassSpec : Spek({
...
+                    assertFindingMessage(findings, message)
+                }
+
+                it("that inherits from an interface") {
+                    val code = """
+                        interface A {
+                            val i: Int
+                        }
+                        abstract class B : A
+                    """.trimIndent()
```
Comment (schalkms) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:70]: I don't think we used the trim statement for the code snippets except for asserts regarding the reported line number.

```suggestion
                    """
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt [hunk 5]
```diff
@@ -38,28 +33,56 @@ class UnnecessaryAbstractClassSpec : Spek({
...
+                }
+
+                it("that inherits from another abstract class") {
+                    val code = """
+                        @Deprecated("We don't care about this first class")
+                        abstract class A {
+                            abstract val i: Int
+                        }
+                        abstract class B : A()
+                    """.trimIndent()
```
Comment (schalkms) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:82]:
```suggestion
                    """
```

**Review by schalkms — COMMENTED** [review]
I guess the mentioned code snippets aren't flagged a second time by the `empty` rule set, right?

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/IndentingXMLStreamWriter.kt (hunk 1 above)
Comment (BraisGabin) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/IndentingXMLStreamWriter.kt:10]: `DelegatingXMLSteamingWriter` was a unnecessary usage of abstract so as soon as I fixed the rule these code was flagged. For that reason it is in the same pr. I can move it to other PR but that one should be merged before this one then.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt (hunk 3 above)
Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt:80]: You are right. I didn't "improve the message". Before there were some issues that had the message "An abstract class without an abstract member can be refactored to a concrete class." but it should be "An abstract class without a concrete member can be refactored to an interface." and the improvement was that. I set the correct message to some issues that were getting the wrong one.

**Comment by BraisGabin** [comment]
> I guess the mentioned code snippets aren't flagged a second time by the `empty` rule set, right?

I guess that some cases are flagged by that rule set too. And I think that it's good that both rules flag it because one developer could have one of those issues disabled.

**Comment by schalkms** [comment]
If I remember correctly, we excluded those cases for exactly this reason. Thus, the same code is not flagged twice by two different rules. I could be wrong here, however.

> I guess that some cases are flagged by that rule set too. And I think that it's good that both rules flag it because one developer could have one of those issues disabled.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt (hunk 3 above)
Comment (schalkms) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt:80]: That's not necessary IMHO.
--> resolved

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/IndentingXMLStreamWriter.kt (hunk 1 above)
Comment (schalkms) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/baseline/IndentingXMLStreamWriter.kt:10]: That's not necessary IMHO.
--> resolved

**Comment by BraisGabin** [comment]
I didn't introduce those cases in this PR. Those cases where there already, they weren't tested but they were flagged. I just extended the tests to ensure that I didn't break anything.

**Comment by schalkms** [comment]
I'm afraid, I didn't understand this statement.  As I understood the description of the rule, you fixed a false-positive in this PR (e.g. `abstract class B : A` as well as `abstract class B : A()`).

> I didn't introduce those cases in this PR. Those cases where there already, they weren't tested but they were flagged.

**Comment by BraisGabin** [comment]
This rule was already flagging code at the same time as the empty rule set. I didn't introduce that issue here. For example:

```kotlin
abstract class A : B {}
```

It was flagged by `UnnecessaryAbstractClass` and by `EmptyClassBlock`.

For that reason I say that I didn't introduce this "double report", it was there. For sure, now there are more "double reports" but that is because now `UnnecessaryAbstractClass` is working as expected.

**Comment by schalkms** [comment]
Thanks for the elaboration, Brais! It's good to merge.

**Review by schalkms — APPROVED** [review]

**Pull request merged**